src/Nethermind/Nethermind.Trie/Pruning/TrieStore.cs

```diff
@@ -21,93 +24,160 @@ namespace Nethermind.Trie.Pruning
+                    SaveInCache(key, trieNode);
                 }
 
                 return trieNode;
             }
 
-            public TrieNode FromCachedRlpOrUnknown(Hash256 hash)
+            public TrieNode FromCachedRlpOrUnknown(in Key key)
             {
                 // ReSharper disable once ConditionIsAlwaysTrueOrFalse
-                if (_objectsCache.TryGetValue(hash, out TrieNode? trieNode))
+                if (TryGetValue(key, out TrieNode trieNode))
                 {
                     if (trieNode!.FullRlp.IsNull)
                     {
                         // // this happens in SyncProgressResolver
                         // throw new InvalidAsynchronousStateException("Read only trie store is trying to read a transient node.");
-                        return new TrieNode(NodeType.Unknown, hash);
+                        return new TrieNode(NodeType.Unknown, key.Keccak);
                     }
 
                     // we returning a copy to avoid multithreaded access
-                    trieNode = new TrieNode(NodeType.Unknown, hash, trieNode.FullRlp);
-                    trieNode.ResolveNode(_trieStore);
-                    trieNode.Keccak = hash;
+                    trieNode = new TrieNode(NodeType.Unknown, key.Keccak, trieNode.FullRlp);
+                    trieNode.ResolveNode(_trieStore.GetTrieStore(key.Address), key.Path);
+                    trieNode.Keccak = key.Keccak;
 
                     Metrics.LoadedFromCacheNodesCount++;
                 }
                 else
                 {
-                    trieNode = new TrieNode(NodeType.Unknown, hash);
+                    trieNode = new TrieNode(NodeType.Unknown, key.Keccak);
                 }
 
                 if (_trieStore._logger.IsTrace) _trieStore._logger.Trace($"Creating new node {trieNode}");
                 return trieNode;
             }
 
-            public bool IsNodeCached(Hash256 hash) => _objectsCache.ContainsKey(hash);
+            private readonly ConcurrentDictionary<Key, TrieNode> _byKeyObjectCache = new();
+            private readonly ConcurrentDictionary<Hash256AsKey, TrieNode> _byHashObjectCache = new();
 
-            public ConcurrentDictionary<Hash256AsKey, TrieNode> AllNodes => _objectsCache;
+            public bool IsNodeCached(in Key key)
+            {
+                if (_storeByHash) return _byHashObjectCache.ContainsKey(key.Keccak);
+                return _byKeyObjectCache.ContainsKey(key);
+            }
 
-            private readonly ConcurrentDictionary<Hash256AsKey, TrieNode> _objectsCache = new();
+            public IEnumerable<KeyValuePair<Key, TrieNode>> AllNodes
+            {
+                get
+                {
+                    if (_storeByHash)
+                    {
+                        return _byHashObjectCache.Select(
+                            pair => new KeyValuePair<Key, TrieNode>(new Key(null, TreePath.Empty, pair.Key.Value), pair.Value));
```

Comment:
why new key needs to be constructed?

Reply:
To match the interface of AllNodes when it is stored by hash, which does not use Key internally.